exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/MetricsRequestMarshaler.java

```diff
@@ -0,0 +1,892 @@
+    }
+
+    private ExemplarMarshaler(
+        long timeUnixNano,
+        long value,
+        int valueFieldNumber,
+        byte[] spanId,
+        byte[] traceId,
+        AttributeMarshaler[] filteredAttributeMarshalers) {
+      super(
+          calculateSize(
+              timeUnixNano, value, valueFieldNumber, spanId, traceId, filteredAttributeMarshalers));
+      this.timeUnixNano = timeUnixNano;
+      this.value = value;
+      this.valueFieldNumber = valueFieldNumber;
+      this.spanId = spanId;
+      this.traceId = traceId;
+      this.filteredAttributeMarshalers = filteredAttributeMarshalers;
+    }
+
+    @Override
+    public void writeTo(CodedOutputStream output) throws IOException {
+      MarshalerUtil.marshalFixed64(
+          io.opentelemetry.proto.metrics.v1.internal.Exemplar.TIME_UNIX_NANO_FIELD_NUMBER,
+          timeUnixNano,
+          output);
+      MarshalerUtil.marshalFixed64(valueFieldNumber, value, output);
+      MarshalerUtil.marshalBytes(
+          io.opentelemetry.proto.metrics.v1.internal.Exemplar.SPAN_ID_FIELD_NUMBER, spanId, output);
+      MarshalerUtil.marshalBytes(
+          io.opentelemetry.proto.metrics.v1.internal.Exemplar.TRACE_ID_FIELD_NUMBER,
+          traceId,
+          output);
+      MarshalerUtil.marshalRepeatedMessage(
+          io.opentelemetry.proto.metrics.v1.internal.Exemplar.FILTERED_ATTRIBUTES_FIELD_NUMBER,
+          filteredAttributeMarshalers,
+          output);
+    }
+
+    private static int calculateSize(
+        long timeUnixNano,
+        long value,
+        int valueFieldNumber,
+        byte[] spanId,
+        byte[] traceId,
+        AttributeMarshaler[] filteredAttributeMarshalers) {
+      int size = 0;
+      size +=
+          MarshalerUtil.sizeFixed64(
+              io.opentelemetry.proto.metrics.v1.internal.Exemplar.TIME_UNIX_NANO_FIELD_NUMBER,
+              timeUnixNano);
+      size += MarshalerUtil.sizeFixed64(valueFieldNumber, value);
+      size +=
+          MarshalerUtil.sizeBytes(
+              io.opentelemetry.proto.metrics.v1.internal.Exemplar.SPAN_ID_FIELD_NUMBER, spanId);
+      size +=
+          MarshalerUtil.sizeBytes(
+              io.opentelemetry.proto.metrics.v1.internal.Exemplar.TRACE_ID_FIELD_NUMBER, traceId);
+      size +=
+          MarshalerUtil.sizeRepeatedMessage(
+              io.opentelemetry.proto.metrics.v1.internal.Exemplar.FILTERED_ATTRIBUTES_FIELD_NUMBER,
+              filteredAttributeMarshalers);
+      return size;
+    }
+  }
+
+  private static int mapToTemporality(
+      io.opentelemetry.sdk.metrics.data.AggregationTemporality temporality) {
+    switch (temporality) {
+      case CUMULATIVE:
+        return AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE_VALUE;
+      case DELTA:
+        return AggregationTemporality.AGGREGATION_TEMPORALITY_DELTA_VALUE;
+    }
+    return AggregationTemporality.AGGREGATION_TEMPORALITY_UNSPECIFIED_VALUE;
+  }
+}
```

Comment:
this class is ginormous. I wonder if we can break it up a bit.

Reply:
Yeah it's probably hairy - should I split in this PR?

Reply:
This one, or a follow-up. I was definitely cross-eyed by the time I got about half way through.

Reply:
Will do in a followup, I don't think it makes sense without a `metrics` subpackage since lots of related classes, but will require a lot of package private -> public transition of existing code.